Fix Perl 5.30 build issues, and remove GIMME_V usage

ychin opened the pull request:
This makes Perl 5.30 dynamic builds work again, by reverting GIMME_V back to GIMME. In Perl headers, GIMME_V requires usage of functions from inline.h which requires manual copy-paste into if_perl.xs which is fragile. Just use GIMME for now (which was the case before) as its implementation in Perl headers does not require using inline functions.

Also, fix the configure script so that if we are using dynamic linkage, we don't pass `-lperl` to the build flags, to avoid accidental external linkage. This is similar to how Python integration works.

Also fix misleading comments in if_perl.xs saying that we include perl\lib\CORE\inline.h manually. That was the case in the past but changed to manually copying inline functions over to fix 5.22 build issues with PERL_NO_INLINE_FUNCTIONS.

ychin committed:
Fix if_perl.xs not being able to build on all versions of Perl (5.30)

This fixes the dynamic builds of Perl interface. The Perl interface file
previously had to manually copy and paste misc inline functions verbatim
from the Perl headers, because we defined `PERL_NO_INLINE_FUNCTIONS`
which prevents us form getting some function definitions. The original
reason we defined it was because those inline functions would reference
Perl functions that would cause linkage errors.

This is a little fragile as every time a new version of Perl comes out,
we inevitably have to copy over new versions of inline functions to our
file, and it's also easy to miss updates to existing functions.

Instead, remove the `PERL_NO_INLINE_FUNCTIONS` define, remove the manual
copy-pasted inline functions. Simply add stub implementations of the
missing linked functions like `Perl_sv_free2` and forward them to the
DLL version of the function at runtime. There are only a few functions
that need this treatment, and it's a simple stub so there is very low
upkeep compared to copying whole implementations to the file.

Also, fix the configure script so that if we are using dynamic linkage,
we don't pass `-lperl` to the build flags, to avoid accidental external
linkage while using dynamic builds. This is similar to how Python
integration works.

k-takata reviewed (approved):
I tested this on Perl 5.32 for dynamic, and it is okay.

k-takata reviewed (approved):
I confirmed with Perl 5.32 and 5.38.

ychin committed:
Fix GIMME_V deprecation warnings in Perl 5.38

Just use GIMME_V, and only use GIMME when using 5.30 to avoid needing to
link Perl_block_gimme. We could provide a stub like the other linked
functions like Perl_sv_free2, but simply using GIMME is the simplest and
it has always worked before.

ychin committed:
Fix Perl 5.38 issues

Fix two issues:

1. Perl 5.38 links against more functions in their inline headers, so we
   need to stub them too.

2. Perl 5.38 made Perl_get_context an inline function, but *only* for
   non-Windows build. Fix that. Note that this was happening in Vim
   currently, as it would build, but fail to run Perl code at runtime.

ychin committed:
Fix Perl 5.36/5.38 when thread local is used

Perl 5.36 introduced using `_Thread_local` for the current context,
which causes inline functions to fail. Create a stub
`PL_current_context` thread local variable to satisfy the linker for
inlined functions. Note that this is going to result in a different
`PL_current_context` being used than the one used in the library, but so
far from testing it seems to work.

ychin committed:
Add docs for how to build Perl for dynamic linking to work
